### PR TITLE
Feature: Add operation collection

### PR DESCRIPTION
**Pull request by fernandes**
Add #collection to operation, so on Controller::index just need to:

```ruby
class Controller < ApplicationController
  def index
    collection Thing::Index
  end
end
```

And a @collection and @things variable will be set with the collection returned by the operation.

To implement, just need to:

```ruby
class Thing < ActiveRecord::Base
  class Index
    def fetch(*params)
  end
end
```

`setup_collection!` is calling `setup_params!` too, if need to make some tweak on the params.

JSON serialization is working perfectly too, just the syntax is a little bit weird (not so compatible with trbrb way) as can see on this [test](https://github.com/apotonick/trailblazer/blob/bf5ca2aac928dab0b853243fe2fb1f166030c19c/test/rails/fake_app/song/operations.rb#L116)

```ruby
class JSON < self
  module BandRepresenter
    include Representable::JSON
    property :name
  end

  self.representer_class = BandRepresenter
end
```

This should be more like the representer DSL but after many hours going deeper on Representable, I could only do so

### Bonus:

When runned the tests under rails with `rake rails` task, many tests were broken, so:
* I [fixed](https://github.com/apotonick/trailblazer/commit/36f87c263e8fa8bd6a6c4fe6861e35e631365729) [them](https://github.com/apotonick/trailblazer/commit/b36525cbcb8aea140d5be487af507462f1270729)
* [removed AR deprecation](https://github.com/apotonick/trailblazer/commit/42b55dd2e52a6b5e8438d263421799963a57c1f6)
* [Added responders gem](https://github.com/apotonick/trailblazer/commit/d2e118d90c7f5d40ce0a149bca736da75256850d) to run tests

### CODE REVIEW

**Comment by paulofabiano** [comment]
Hi, @fernandes!
I don't know what's going on but it is not working for me. I implemented the operation call with collection and also created the operation inheriting from my create Operation with fetch method with my User.all. Also updated the gem directly from the git repo. But i'm getting this:

undefined method `collection' for #<Admin::UsersController:0x007f99affe7db0>

Any thoughts?

**Comment by fernandes** [comment]
@paulofabiano its not merged yet... its just a PR suggestion how to solve the Index operation problem, if you wanna test use:

gem 'trailblazer', github: 'fernandes/trailblazer', branch: 'feature/add_operation_collection'

but don't blame if its not merged in the future hahaha

**Comment by apotonick** [comment]
Thanks baby, but I do not entirely understand the rational behind this PR.

I can see the addition to the controller where the collection is stored in `@collection`. However, everything else could be achieved by overriding `#setup_model!´, right? 

Here's your code.

```ruby
class Index
  def fetch(*params)
    Song.all
  end
end
```

I would simply do this.

```ruby
class Index
  def setup_model!(*params)
    Song.all
  end
end
```
Or am I missing anything here? My point is that we add loads of code (and additional API for `Operation` and `Controller` for a rather little change of behaviour, and you know how much I fear code :wink:!

Thanks for your work @fernandes darling! :heart:

**Comment by fernandes** [comment]
1. `collection` is the foundation for scope and pagination, in https://github.com/apotonick/trailblazer/pull/40 you have the big picture overview...

2. `collection` has some different behavior compared to present, like performing search, pagination AND setting variable like @things in the view... so its better to keep compatibility with other stuff

**Comment by fernandes** [comment]
just keeping https://github.com/apotonick/trailblazer/pull/40 once its include everything here

File: lib/trailblazer/operation/controller.rb [hunk 1]
```diff
@@ -70,14 +81,25 @@ def operation!(operation_class, params)
...
 
   def setup_operation_instance_variables!
     @form = @operation.contract
     @model = @operation.model
   end
+  
+  def setup_operation_collection_variables!
+    @collection = @operation.collection
+    operation_collection_name = @operation.collection.model.table_name
+    instance_variable_set(:"@#{operation_collection_name}", @collection)
```
Comment (fernandes) [inline lib/trailblazer/operation/controller.rb]: I'm gonna put this on https://github.com/apotonick/trailblazer/blob/master/lib/trailblazer/operation/crud.rb

File: lib/trailblazer/operation.rb [hunk 2]
```diff
@@ -45,6 +45,18 @@ def present(*params)
       def contract(&block)
         contract_class.class_eval(&block)
       end
+      
+      def collection(*params, &block)
```
Comment (apotonick) [inline lib/trailblazer/operation.rb]: I like the idea of this, and the tests. :kiss: 

However, there's no need to introduce a new API. What if we just add an `Operation::Collection` module that overrides a few methods?

`#model!` and `#setup_model!` are exactly designed to do what your `#fetch` does. This way, we can use `Op::present` without changing a single line of code (maybe skip creating the form object, though).

Not sure if we even need the `Controller#collection` method - it shouldn't matter to the invoker whether you're dealing with a collection or a model.

Comment (fighella) [inline lib/trailblazer/operation.rb]: A personal favourite feature.

Comment (fernandes) [inline lib/trailblazer/operation.rb]: Benefits of this API:

1. We can handle in different ways a single model and collection (duh!), so we don't need to handle using `ifs` each case
2. In collections we can set [scope](https://github.com/apotonick/trailblazer/pull/40/files#diff-c518d16668a2dc2a2951d3206f055936) and [pagination](https://github.com/apotonick/trailblazer/pull/40/files#diff-9355b6bc2d6c54ad4882dcf23cbff87a) out-of-the-box, then we just need to include [Scope](https://github.com/apotonick/trailblazer/pull/40/files#diff-6e0b5532e3691435735d5f7ebe67843eR7) and [Pagination](https://github.com/apotonick/trailblazer/pull/40/files#diff-0cfc7048e88372dae0e3076e477ef5e0R7) (links are for tests) on Collection operations
3. in crud.rb we can set [collection variables](https://github.com/apotonick/trailblazer/pull/36/files#diff-df1b16d18715ddbab4a106cee89a9b4cR100)

**Comment by fernandes** [comment]
@paulofabiano @fighella `collection` in on the master right now :)

controller:
```ruby
def index
  collection Thing::Index
end
```

```ruby
class Thing::Index < Trailblazer::Operation
  include Collection

  def model!(params)
    Thing.all
  end
end
```

access `@collection` on your views

**Pull request closed without merging**